akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala

```diff
@@ -124,10 +135,29 @@ import akka.stream.scaladsl.Source
         .mapMaterializedValue(_ => NotUsed)
 
     val handlerFlow: Flow[(Offset, Envelope), Offset, NotUsed] =
-      Flow[(Offset, Envelope)].mapAsync(parallelism = 1) {
-        case (offset, envelope) =>
-          applyUserRecovery(handler, envelope, offset, logger, () => handler.process(envelope))
-            .map(_ => offset)
+      handlerStrategy match {
+        case grouped: GroupedHandlerStrategy[Envelope] =>
+          Flow[(Offset, Envelope)]
+            .groupedWithin(grouped.afterEnvelopes, grouped.orAfterDuration)
+            .filterNot(_.isEmpty)
+            .mapAsync(parallelism = 1) { group =>
+              val firstOffset = group.head._1
+              val lastOffset = group.last._1
+              val envelopes = group.map(_._2)
+              applyUserRecovery[Offset, immutable.Seq[Envelope]](
+                grouped.handler,
+                envelopes,
+                firstOffset,
```

Comment:
This will be confusing, no? The log message inside `applyUserRecovery` will always mention the first offset. 

We could eventually pass a `Seq[Offset]` and adapt the logging if we have one or many offsets.

Reply:
ok, I can change it to pass in firstOffset and lastOffset and log accordingly

Reply:
Can we then pass a `Seq(singleOffset)` or `Seq(firstOffset, lastOffset)`?

Reply:
I'd prefer to not create an additional Seq instance, since this called for each env